src/cmd/services/m3comparator/main/querier_test.go

```diff
@@ -0,0 +1,218 @@
+	return h.iters, nil
+}
+
+var _ seriesLoadHandler = (*testSeriesLoadHandler)(nil)
+
+type tagMap map[string]string
+
+var (
+	iteratorOpts = iteratorOptions{
+		encoderPool:   encoderPool,
+		iteratorPools: iterPools,
+		tagOptions:    tagOptions,
+		iOpts:         iOpts,
+	}
+	metricNameTag = string(iteratorOpts.tagOptions.MetricName())
+)
+
+func TestFetchCompressedReturnsPreloadedData(t *testing.T) {
+	ctrl := xtest.NewController(t)
+	defer ctrl.Finish()
+
+	predefinedSeriesCount := 100
+	iters := encoding.NewMockSeriesIterators(ctrl)
+	iters.EXPECT().Len().Return(predefinedSeriesCount).MinTimes(1)
+	iters.EXPECT().Close()
+
+	seriesLoader := &testSeriesLoadHandler{iters}
+
+	querier := &querier{opts: iteratorOpts, handler: seriesLoader}
+
+	query := matcherQuery(t, metricNameTag, "preloaded")
+
+	result, cleanup, err := querier.FetchCompressed(nil, query, nil)
+	assert.NoError(t, err)
+	defer cleanup()
+
+	assert.Equal(t, predefinedSeriesCount, result.SeriesIterators.Len())
+}
+
+func TestFetchCompressedGeneratesRandomData(t *testing.T) {
+	tests := []struct {
+		name       string
+		givenQuery *storage.FetchQuery
+		wantSeries []tagMap
+	}{
+		{
+			name:       "random data for known metrics",
+			givenQuery: matcherQuery(t, metricNameTag, "quail"),
+			wantSeries: []tagMap{
+				{
+					metricNameTag: "quail",
+					"foobar":      "qux",
+					"name":        "quail",
+				},
+			},
+		},
+		{
+			name:       "a hardcoded list of metrics",
+			givenQuery: matcherQuery(t, metricNameTag, "unknown"),
+			wantSeries: []tagMap{
+				{
+					metricNameTag: "foo",
+					"foobar":      "qux",
+					"name":        "foo",
+				},
+				{
+					metricNameTag: "bar",
+					"foobar":      "qux",
+					"name":        "bar",
+				},
+				{
+					metricNameTag: "quail",
+					"foobar":      "qux",
+					"name":        "quail",
+				},
+			},
+		},
+		{
+			name:       "a given number of single series metrics",
+			givenQuery: matcherQuery(t, "gen", "2"),
+			wantSeries: []tagMap{
+				{
+					metricNameTag: "foo_0",
+					"foobar":      "qux",
+					"name":        "foo_0",
+				},
+				{
+					metricNameTag: "foo_1",
+					"foobar":      "qux",
+					"name":        "foo_1",
+				},
+			},
+		},
+		{
+			name:       "single metrics with a given number of series",
+			givenQuery: matcherQuery(t, metricNameTag, "multi_4"),
+			wantSeries: []tagMap{
+				{
+					metricNameTag: "multi_4",
+					"const":       "x",
+					"id":          "0",
+					"parity":      "0",
+				},
+				{
+					metricNameTag: "multi_4",
+					"const":       "x",
+					"id":          "1",
+					"parity":      "1",
+				},
+				{
+					metricNameTag: "multi_4",
+					"const":       "x",
+					"id":          "2",
+					"parity":      "0",
+				},
+				{
+					metricNameTag: "multi_4",
+					"const":       "x",
+					"id":          "3",
+					"parity":      "1",
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
```

Comment:
nit: use `xtest.NewController(t)` here as well, it's a wrapper on `gomock.NewController` that doesn't hang tests when an unexepcted method is called